m3/src/reporter_impl.cc

```diff
@@ -0,0 +1,438 @@
+                                           int precision) {
+  // Format the duration into the fraction of the given precision (i.e v /
+  // (10^p)). Omit any trailing zeros and the decimal when there is no
+  // fractional part.
+  double numerator = static_cast<double>(duration.count());
+  double divisor = std::pow(10, precision);
+  std::string str = std::to_string(numerator / divisor);
+
+  // Remove trailing zeros.
+  while (str[str.size() - 1] == '0') {
+    str.erase(str.size() - 1);
+  }
+
+  // Remove the decimal if it is the last character.
+  if (str[str.size() - 1] == '.') {
+    str.erase(str.size() - 1);
+  }
+
+  return str;
+}
+
+void Reporter::Impl::Run() {
+  std::shared_ptr<TCalcTransport> calc_transport(new TCalcTransport());
+
+  std::unique_lock<std::mutex> run_lock(run_mutex_);
+  while (run_) {
+    cv_.wait(run_lock);
+    if (!run_) {
+      // When the reporter is closed, this thread needs to drain the queue and
+      // flush any buffered metrics before it exits.
+      std::lock_guard<std::mutex> queue_lock(queue_mutex_);
+      while (queue_.size() > 0) {
+        auto metric = queue_.front();
+        Process(metric, calc_transport);
+        queue_.pop();
+      }
+      Flush();
+      return;
+    }
+
+    // While the queue is not empty this thread will continue to dequeue metrics
+    // without waiting on the condition variable.
+    while (true) {
+      std::unique_lock<std::mutex> queue_lock(queue_mutex_);
+      if (queue_.size() == 0) {
+        break;
+      }
+      auto metric = queue_.front();
+
+      // Release the queue lock so other threads can enqueue metrics while this
+      // thread processes its current batch of metrics.
+      queue_lock.unlock();
+
+      Process(metric, calc_transport);
+
+      queue_lock.lock();
+      queue_.pop();
+    }
+  }
+}
+
+void Reporter::Impl::Process(thrift::Metric metric,
+                             std::shared_ptr<TCalcTransport> calc_transport) {
+  TCompactProtocol calc_protocol(calc_transport);
+  metric.write(&calc_protocol);
+  auto size = calc_transport->size();
+
+  if (size > max_packet_size_) {
+    Flush();
+    calc_transport->flush();
+
+    thrift::MetricBatch batch;
+    batch.__set_commonTags(common_tags_);
+    batch.write(&calc_protocol);
+
+    // The current metric is not included in the batch of metrics that has just
+    // been flushed so we need to re-add it's size after the reset.
+    metric.write(&calc_protocol);
+  }
+
+  std::lock_guard<std::mutex> lock(emission_mutex_);
+  emission_batch_.metrics.push_back(metric);
+}
+
+void Reporter::Impl::Flush() {
+  std::lock_guard<std::mutex> lock(emission_mutex_);
+
+  if (emission_batch_.metrics.empty()) {
+    return;
+  }
+
+  try {
+    emission_client_->emitMetricBatch(emission_batch_);
+  } catch (const TTransportException &e) {
+    std::cerr << "Encountered error emitting M3 metric batch: " << e.what()
```

Comment:
hm could we use a levelled logger which includes filename/function-name/linenumber/etc instead of emitting straight to stderr

Reply:
I don't have strong feelings here, the ATG code I looked at logs to stderr so I was following the same pattern, we can sync with them.

Reply:
Sounds good.